README.md
# Revision Strike

[![Build Status](https://travis-ci.org/stevegrunwell/revision-strike.svg?branch=master)](https://travis-ci.org/stevegrunwell/revision-strike)
[![Code Climate](https://codeclimate.com/github/stevegrunwell/revision-strike/badges/gpa.svg)](https://codeclimate.com/github/stevegrunwell/revision-strike)
[![Test Coverage](https://codeclimate.com/github/stevegrunwell/revision-strike/badges/coverage.svg)](https://codeclimate.com/github/stevegrunwell/revision-strike/coverage)

Unless post revisions are explicitly limited, WordPress will build up a hefty sum of revisions over time. While it's great to have revision history for some recent content, the chances that old revisions will be necessary diminish the longer a post has been published. Revision Strike is designed to automatically remove these unneeded revisions on older, published posts.

**How does it work?**

First, a threshold is set, with a default of 30 days. Once a day, Revision Strike will run and find any post revisions in the database attached to **published** posts with a post date of at least 30 (or your custom threshold) days ago, and "strike" (tear-down and remove) them from the WordPress database.


## Usage

There are a number of ways to interact with Revision Strike:


### WP Cron

Upon plugin activation, a hook is registered to trigger the `revisionstrike_strike_old_revisions` action daily, which kicks off the striking process. This hook is then automatically removed upon plugin deactivation.

The WP Cron method is great for ongoing maintenance, but when you manually need to purge larger amounts of revisions&hellip;


### Tools &rsaquo; Revision Strike

![The Tools > Revision Strike page](plugin-repo-assets/screenshot-1.png)

You can manually trigger a revision strike by logging into WordPress and visiting the Tools &rsaquo; Revision Strike page.

> **Note:** You must have the "edit_others_posts" capability (typically "Editor"-level and above) in order to see this page.


### WP-CLI

If you make use of [WP-CLI](http://wp-cli.org/) on your site you may trigger Revision Strike with the following commands:

```bash
$ wp revision-strike <command>
```


#### clean

Manually runs an instance of Revision Strike, removing revisions matching the passed arguments or defaulting to the values declared on the Settings &rsaquo; Writing page.

##### Arguments

<dl>
	<dt>--days=&lt;days&gt;</dt>
	<dd>Remove revisions on posts published at least &lt;days&gt; day(s) ago. This is determined by the value set on Settings &rsaquo; Writing or a default of 30.</dd>
	<dt>--limit=&lt;limit&gt;</dt>
	<dd>The maximum number of revisions to remove. This is determined by the value set on Settings &rsaquo; Writing or a default value of 50.</dd>
	<dt>--post_type=&lt;post_type&gt;</dt>
	<dd>One or more post types (comma-separated) for which revisions should be struck. Default value is 'post'.</dd>
	<dt>--verbose</dt>
	<dd>Enable verbose logging of deleted revisions.</dd>
</dl>


#### clean-all

If you have a large number of revisions, it can take Revision Strike some time to work its way through your post history to clean up revisions on a daily basis. The `clean-all` command is like spring cleaning, systematically removing eligible post revisions all at once, giving you a nice, clean revision history to start from, making it easy for the daily cron runner to maintain.

Under the hood, this command just determines how many eligible revisions are in the database, then calls the `clean` command with your arguments and a limit equal to the number of revisions in the database. Thanks to the revision chunking ([#17](https://github.com/stevegrunwell/revision-strike/issues/17)), no more than 50 post revisions are being handled in memory at any given time, ensuring minimal performance impact on your site.

##### Arguments

<dl>
	<dt>--days=&lt;days&gt;</dt>
	<dd>Remove revisions on posts published at least &lt;days&gt; day(s) ago. This is determined by the value set on Settings &rsaquo; Writing or a default of 30.</dd>
	<dt>--post_type=&lt;post_type&gt;</dt>
	<dd>One or more post types (comma-separated) for which revisions should be struck. Default value is 'post'.</dd>
	<dt>--verbose</dt>
	<dd>Enable verbose logging of deleted revisions.</dd>
</dl>


## Hooks

Revision Strike leverages the [WordPress Plugin API](https://codex.wordpress.org/Plugin_API) to let you customize its behavior without editing the codebase directly.

### Filter: `revisionstrike_post_types`

Controls the post types for which revisions should be automatically be purged.

<dl>
	<dt>(string) <code>$post_types</code></dt>
	<dd>A comma-separated list of post types.</dd>
</dl>

#### Example

By default, Revision Strike only works against posts of the "Post" post type. If you'd like to include other post types (pages, custom post types, etc.), this filter will allow you do so.

```php
/**
 * Include the "page" and "book" custom post type in Revision Strike.
 *
 * @param string $post_type A comma-separated list of post types.
 */
function theme_set_post_types( $post_types ) {
	return 'post,page,book';
}
add_filter( 'revisionstrike_post_types', 'theme_set_post_types' );
```

### Filter: `revisionstrike_get_revision_ids`

Filter the list of eligible revision IDs.

<dl>
	<dt>(array) <code>$revision_ids</code></dt>
	<dd>Revision IDs to be struck.</dd>
	<dt>(int) <code>$days</code></dt>
	<dd>The number of days since a post's publish date that must pass before we can purge the post revisions.</dd>
	<dt>(int) <code>$limit</code></dt>
	<dd>The maximum number of revision IDs to retrieve.</dd>
	<dt>(array) <code>$post_type</code></dt>
	<dd>The post types for which revisions should be located.</dd>
</dl>

### Filter: `revisionstrike_capabilities`

Sets capabilities to access Revision Strike managment page.

#### Example

By default, Revision Strike sets required capabilities to access plugin managment page as `edit_others_posts`. You can use this filter to change this in order to have more control on who can access it.

```php
/**
 * Sets capabilities to access Revision Strike settings page.
 */
function rs_capabilities() {
	return 'manage_options';
}
add_filter( 'revisionstrike_capabilities', 'rs_capabilities' );
```


## Contributing

All development dependencies for the plugin are installed via [Composer](https://getcomposer.org/). After installing Composer on your system, navigate your terminal session to the plugin directory and run:

	$ composer install

Pull requests on this plugin are welcome, but I ask that you please follow these guidelines:

* Every PR that touches code should include a corresponding test. This plugin uses [PHPUnit](https://phpunit.de/), [WP_Mock](https://github.com/10up/wp_mock), [Mockery](http://docs.mockery.io/en/latest/), and [Behat](http://behat.org). Every PR should [generate a passing build in Travis CI](https://travis-ci.org/stevegrunwell/revision-strike) before being considered for merger.
* Please follow the [WordPress Coding Standards](https://codex.wordpress.org/WordPress_Coding_Standards). If you have any doubts, you can run your PHP through [PHP_CodeSniffer](https://www.squizlabs.com/php-codesniffer) using the **RevisionStrike** ruleset:

		$ vendor/bin/phpcs

	This CodeSniffer ruleset combines the "WordPress-Extra" and "WordPress-Docs" rulesets for a *very* specific level of coverage; beyond the coding standards, contributed inline documentation will be checked against the [WordPress Inline Documentation Standards](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/).

### Integration tests

There are a few integration tests for the plugin, written using [Behat](http://behat.org) and scaffolded using [`wp scaffold package-tests`](http://wp-cli.org/commands/scaffold/package-tests/). If you'd like to contribute to these tests, please install the WP CLI Testing framework:

```bash
$ bin/install-package-tests.sh
$ vendor/bin/behat
```


CHANGELOG.md
# Revision Strike Change Log

All notable changes to this project will be documented in this file, according to [the Keep a Changelog standards](http://keepachangelog.com/).

This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

* Fixed bug where warnings were being thrown when Revision Strike was being run verbosely via WP-CLI. Props @ivankruchkoff. ([#29])
* Re-licensed the plugin under the MIT license.
* Updated Composer dependencies and bumped the minimum development version of PHP to 7.0 (though the plugin is still targeting PHP 5.3+). ([#32])


## [0.5.0] - 2017-05-15

* Add the `revisionstrike_capabilities` filter to allow setting required capabilities to access Revision Strike settings page. Props @pereirinha. ([#28])
* Updated Composer and npm dependencies.
* Update the Code Climate configuration.
* Coding standards cleanup.


## [0.4.1] - 2017-03-04

* Converted unit tests to use the 1.0.x development branch of [WP_Mock](https://github.com/10up/wp_mock)
* Adjusted visibility of `RevisionStrikeCLI::log_deleted_revision()`. ([#20])


## [0.4.0] - 2016-07-29

* Move from the manual pre-commit hook to [WP Enforcer](https://github.com/stevegrunwell/wp-enforcer).
	* As a result, minor standards-related changes have been made to the code to comply with the [WordPress Coding Standards](https://codex.wordpress.org/WordPress_Coding_Standards).
* Re-work the copy on Tools &rsaquo; Revision Strike to be more clear to site administrators. ([#26])
* Add additional tests around the Tools &rsaquo; Revision Strike page confirmation messages. ([#25])


## [0.3.0] - 2016-06-20

* Lock Composer dependency versions to ensure more consistent testing via Travis-CI.
* Add the `revisionstrike_get_revision_ids` filter to enable third-party plugins and themes to alter the array of revision IDs. ([#21])
* Implement Grunt to more consistently build releases. ([#18])


## [0.2.0] - 2015-08-16

* Added a "Limit" setting to Settings &rsaquo; Writing. ([#13])
* Added a "clean-all" WP-CLI command. ([#14])
* Clarified language on the Settings &rsaquo; Writing and Tools &rsaquo; Revision Strike pages. Props to @GhostToast for the suggestion! ([#16])
* Strike requests are now batched into groupings of 50 IDs at a time to avoid overwhelming underpowered machines. ([#17])


## [0.1.0] - 2015-08-09

Initial public release.


[Unreleased]: https://github.com/stevegrunwell/revision-strike/compare/master...develop
[0.5.0]: https://github.com/stevegrunwell/revision-strike/compare/v0.4.1...v0.5.0
[0.4.1]: https://github.com/stevegrunwell/revision-strike/compare/v0.3.0...v0.4.1
[0.4.0]: https://github.com/stevegrunwell/revision-strike/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/stevegrunwell/revision-strike/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/stevegrunwell/revision-strike/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/stevegrunwell/revision-strike/releases/tag/v0.1.0
[#13]: https://github.com/stevegrunwell/revision-strike/issues/13
[#14]: https://github.com/stevegrunwell/revision-strike/issues/14
[#16]: https://github.com/stevegrunwell/revision-strike/issues/16
[#17]: https://github.com/stevegrunwell/revision-strike/issues/17
[#18]: https://github.com/stevegrunwell/revision-strike/issues/18
[#20]: https://github.com/stevegrunwell/revision-strike/issues/20
[#21]: https://github.com/stevegrunwell/revision-strike/issues/21
[#25]: https://github.com/stevegrunwell/revision-strike/issues/25
[#26]: https://github.com/stevegrunwell/revision-strike/issues/26
[#28]: https://github.com/stevegrunwell/revision-strike/pull/28
[#29]: https://github.com/stevegrunwell/revision-strike/pull/29
[#32]: https://github.com/stevegrunwell/revision-strike/pull/32


tests/PHPUnit/CLITest.php
<?php
/**
 * Tests for the plugin's bootstrapping.
 *
 * @package Revision Strike
 * @author Steve Grunwell
 */

namespace Grunwell\RevisionStrike;

use WP_Mock as M;
use Mockery;
use ReflectionMethod;
use ReflectionProperty;
use RevisionStrike;
use RevisionStrikeCLI;
use WP_CLI;

class CLITest extends TestCase {

	protected $testFiles = [
		'class-revision-strike.php',
		'class-revision-strike-cli.php',
		'class-settings.php',
	];

	public function tearDown() {
		parent::tearDown();

		\WP_CLI::tearDown();
	}

	public function test_clean() {
		$wp_cli = WP_CLI::getInstance();
		$wp_cli->shouldReceive( '_line' )->once();
		$wp_cli->shouldReceive( '_success' )->once();

		$instance = Mockery::mock( 'RevisionStrike' )->makePartial();
		$instance->shouldReceive( 'strike' )->once();
		$instance->shouldReceive( 'get_stats' )
			->once()
			->andReturn( array( 'deleted' => 50, ) );

		$cli = Mockery::mock( 'RevisionStrikeCLI' )
			->shouldAllowMockingProtectedMethods()
			->makePartial();
		$cli->shouldReceive( 'get_instance' )
			->once()
			->andReturn( $instance );

		M::passthruFunction( '_n' );

		$cli->clean( array(), array() );
	}

	public function test_clean_with_days_argument() {
		$instance = Mockery::mock( 'RevisionStrike' )->makePartial();
		$instance->shouldReceive( 'strike' )
			->once()
			->with( array( 'days' => 7, ) );
		$instance->shouldReceive( 'get_stats' )
			->once()
			->andReturn( array( 'deleted' => 0, ) );

		$cli = Mockery::mock( 'RevisionStrikeCLI' )
			->shouldAllowMockingProtectedMethods()
			->makePartial();
		$cli->shouldReceive( 'get_instance' )
			->once()
			->andReturn( $instance );

		M::passthruFunction( 'esc_html__' );

		$cli->clean( array(), array( 'days' => 7 ) );
	}

	public function test_clean_with_limit_argument() {
		$instance = Mockery::mock( 'RevisionStrike' )->makePartial();
		$instance->shouldReceive( 'strike' )
			->once()
			->with( array( 'limit' => 100, ) );
		$instance->shouldReceive( 'get_stats' )
			->once()
			->andReturn( array( 'deleted' => 0, ) );

		$cli = Mockery::mock( 'RevisionStrikeCLI' )
			->shouldAllowMockingProtectedMethods()
			->makePartial();
		$cli->shouldReceive( 'get_instance' )
			->once()
			->andReturn( $instance );

		M::passthruFunction( 'esc_html__' );

		$cli->clean( array(), array( 'limit' => 100 ) );
	}

	public function test_clean_with_post_type_argument() {
		$instance = Mockery::mock( 'RevisionStrike' )->makePartial();
		$instance->shouldReceive( 'strike' )
			->once()
			->with( array( 'post_type' => 'page', ) );
		$instance->shouldReceive( 'get_stats' )
			->once()
			->andReturn( array( 'deleted' => 0 ) );

		$cli = Mockery::mock( 'RevisionStrikeCLI' )
			->shouldAllowMockingProtectedMethods()
			->makePartial();
		$cli->shouldReceive( 'get_instance' )
			->once()
			->andReturn( $instance );

		M::passthruFunction( 'esc_html__' );

		$cli->clean( array(), array( 'post_type' => 'page' ) );
	}

	public function test_clean_with_verbose_argument() {
		$this->markTestSkipped( 'Tests need refactoring, and the version of WP_Mock currently being used isn\'t working properly with closures' );
		$instance = Mockery::mock( 'RevisionStrike' )->makePartial();
		$instance->shouldReceive( 'strike' )->once();
		$instance->shouldReceive( 'get_stats' )
			->once()
			->andReturn( array( 'deleted' => 0, ) );

		$cli = Mockery::mock( 'RevisionStrikeCLI' )
			->shouldAllowMockingProtectedMethods()
			->makePartial();
		$cli->shouldReceive( 'get_instance' )
			->once()
			->andReturn( $instance );

		M::passthruFunction( 'esc_html__' );

		M::expectActionAdded( 'wp_delete_post_revision', M\Functions::type('callable'));

		$cli->clean( array(), array( 'verbose' => true ) );
	}

	public function test_clean_reporting() {
		$instance = Mockery::mock( 'RevisionStrike' )->makePartial();
		$instance->shouldReceive( 'strike' )->once();
		$instance->shouldReceive( 'get_stats' )
			->once()
			->andReturn( array( 'deleted' => 50, ) );

		$cli = Mockery::mock( 'RevisionStrikeCLI' )
			->shouldAllowMockingProtectedMethods()
			->makePartial();
		$cli->shouldReceive( 'get_instance' )
			->once()
			->andReturn( $instance );

		M::passthruFunction( 'esc_html__', array(
			'times' => 0,
		) );
		M::passthruFunction( '_n' );

		$cli->clean( array(), array() );
	}

	public function test_clean_all() {
		$wp_cli = WP_CLI::getInstance();

		$instance = Mockery::mock( 'RevisionStrike' )->makePartial();
		$instance->shouldReceive( 'count_eligible_revisions' )
			->once()
			->with( 45, 'post,page' )
			->andReturn( 777 );

		$settings = Mockery::mock( 'RevisionStrikeSettings' )->makePartial();
		$settings->shouldReceive( 'get_option' )
			->never()
			->with( 'days' );
		$settings->shouldReceive( 'get_option' )
			->once()
			->with( 'post_type' )
			->andReturn( 'post,page' );
		$instance->settings = $settings;

		$cli = Mockery::mock( 'RevisionStrikeCLI' )
			->shouldAllowMockingProtectedMethods()
			->makePartial();
		$cli->shouldReceive( 'get_instance' )
			->once()
			->andReturn( $instance );
		$cli->shouldReceive( 'clean' )
			->once()
			->with( array(), array( 'days' => 45, 'limit' => 777 ) );

		$cli->clean_all( array(), array( 'days' => 45, ) );
	}

	public function test_get_instance() {
		$instance = new RevisionStrikeCLI;
		$method   = new ReflectionMethod( $instance, 'get_instance' );
		$method->setAccessible( true );
		$property = new ReflectionProperty( $instance, 'instance' );
		$property->setAccessible( true );

		$this->assertNull( $property->getValue( $instance ) );

		$method->invoke( $instance );

		$this->assertInstanceOf( 'RevisionStrike', $property->getValue( $instance ) );
	}

}

tests/PHPUnit/ToolsTest.php
<?php
/**
 * Tests for the Tools > Revision Strike page.
 *
 * @package Revision Strike
 * @author Steve Grunwell
 */

namespace Grunwell\RevisionStrike;

use WP_Mock as M;
use Mockery;
use ReflectionMethod;
use ReflectionProperty;
use RevisionStrike;

class ToolsTest extends TestCase {

	protected $tools_file;

	public function setup() {
		parent::setup();

		$this->tools_file = PROJECT . 'tools.php';

		M::passthruFunction( '__' );
		M::passthruFunction( '_x' );
		M::passthruFunction( 'absint' );
		M::passthruFunction( 'esc_attr' );
		M::passthruFunction( 'esc_html' );
		M::passthruFunction( 'esc_html__' );
		M::passthruFunction( 'esc_html_e' );
		M::passthruFunction( 'esc_url' );
		M::passthruFunction( 'wp_kses_post' );
	}

	public function test_default_load() {
		M::userFunction( 'wp_nonce_url', array(
			'times'  => 1,
			'args'   => array( 'tools.php?page=revision-strike', 'revision-strike', 'nonce' ),
		) );

		M::userFunction( 'admin_url', array(
			'times'  => 1,
			'args'   => array( 'options-writing.php' ),
		) );

		M::userFunction( 'submit_button', array(
			'times'  => 1,
		) );

		$instance = new \stdClass;
		$defaults = array(
			'days'  => 15,
			'limit' => 50,
		);

		ob_start();
		include $this->tools_file;
		$results = ob_get_contents();
		ob_end_clean();

		$this->assertContains(
			'<input name="days" type="number" class="small-text" value="15" />',
			$results,
			'Days field is not being populated'
		);
		$this->assertContains(
			'<input name="limit" type="number" class="small-text" value="50" />',
			$results,
			'Limit field is not being populated'
		);
	}

	public function test_load_when_defaults_array_is_undefined() {
		M::userFunction( 'wp_die', array(
			'times'  => 1,
		) );

		ob_start();
		include $this->tools_file;
		ob_end_clean();
	}

	public function test_load_with_valid_nonce() {
		$defaults     = array(
			'days'  => 30,
			'limit' => 50,
		);
		$_GET     = array( 'nonce' => 'MYNONCE' );
		$_POST    = $defaults;
		$instance = Mockery::mock( 'RevisionStrike' )->makePartial();
		$instance->shouldReceive( 'strike' )
			->once()
			->with( $defaults );
		$instance->shouldReceive( 'get_stats' )
			->once()
			->andReturn( array( 'count' => 0 ) );

		M::userFunction( 'wp_verify_nonce', array(
			'times'  => 1,
			'args'   => array( 'MYNONCE', 'revision-strike' ),
			'return' => true,
		) );

		ob_start();
		include $this->tools_file;
		ob_end_clean();

		unset( $_GET, $_POST );
	}

	public function test_load_message_scenarios() {
		$defaults     = array(
			'days'  => 30,
			'limit' => 50,
		);
		$_GET     = array( 'nonce' => true );
		$_POST    = $defaults;
		$instance = Mockery::mock( 'RevisionStrike' )->makePartial();
		$instance->shouldReceive( 'strike' )
			->with( $defaults );
		$instance->shouldReceive( 'get_stats' )
			->times( 3 )
			->andReturn(
				array( 'count' => 0 ), // No revisions found
				array( 'count' => 5, 'deleted' => 0 ), // Something went wrong
				array( 'count' => 5, 'deleted' => 5 ) // We're good
			);

		M::userFunction( 'wp_verify_nonce', array(
			'return' => true,
		) );

		M::userFunction( '_n', array(
			'times'  => 1,
			'return' => '%d post revisions',
		) );

		M::passthruFunction( 'wp_nonce_url' );
		M::passthruFunction( 'submit_button' );

		// Pass 1: no revisions found
		ob_start();
		include $this->tools_file;
		$result = ob_get_contents();
		ob_end_clean();
		$this->assertContains( '<div class="error"><p>', $result, 'No revisions found did not trigger an error message' );

		// Pass 2: revisions found, but nothing deleted
		ob_start();
		include $this->tools_file;
		$result = ob_get_contents();
		ob_end_clean();
		$this->assertContains( '<div class="error"><p>', $result, 'The user should be warned when revisions were found be we can\'t delete them' );

		// Pass 3: revisions deleted successfully
		ob_start();
		include $this->tools_file;
		$result = ob_get_contents();
		ob_end_clean();
		$this->assertContains( '<div class="updated"><p>', $result, 'Notify the user when revisions are deleted' );
		$this->assertContains( '5 post revisions', $result, 'Display the number of revisions that were removed.' );

		unset( $_GET, $_POST );
	}

}

tests/dummy-files/wp-cli.php
<?php
/**
 * Mock class for WP-CLI.
 *
 * Sadly, WP_CLI uses a lot of public, static methods for things like logging. Great when you're
 * writing the code, but a real pain in the butt to test.
 *
 * This test-double for WP_CLI uses the __callStatic magic method paired with a Singleton pattern
 * to effectively convert static method calls to non-static ones (albeit with a "_" prefix).
 *
 * Calling WP_CLI::getInstance() will return a Mockery-powered mock of the class, to which you can
 * attach assertions. Be sure to write your assertions based on the method *with a leading
 * underscore!*
 *
 * Example:
 *
 * public function testHelloWorld() {
 *   $cli = WP_CLI::getInstance();
 *   $cli->shouldReceive( '_success' )
 *     ->once()
 *     ->with( 'Hello world!' );
 * }
 *
 * Be sure to call WP_CLI::tearDown() to reset the Singleton after each test to prevent your tests
 * from getting polluted with old data!
 */

if ( class_exists( '\WP_CLI' ) || class_exists( '\WP_CLI_Command' ) ) {
	return;
}

class WP_CLI {

	public static $_instance;

	public static function __callStatic( $name, $args ) {
		$instance = self::getInstance();
		return call_user_func_array( array( $instance, 'calledStatic' ), func_get_args() );
	}

	public static function getInstance() {
		if ( null === static::$_instance ) {
			static::$_instance = \Mockery::mock( __CLASS__ )
				->shouldAllowMockingProtectedMethods()
				->makePartial();
		}
		return static::$_instance;
	}

	/**
	 * Here's something you don't often see in a Singleton: something to destroy it!
	 */
	public static function tearDown() {
		static::$_instance = null;
	}

	public function calledStatic( $name, $args ) {
		$method = '_' . $name;
		$this->$method( $args );
	}

	protected function _add_command() {}
	protected function _line() {}
	protected function _log() {}
	protected function _success() {}
	protected function _error() {}

}

class WP_CLI_Command {

}

includes/tools.php
<?php
/**
 * The Tools > Revision Strike page.
 *
 * @package Revision Strike
 * @author  Steve Grunwell
 */

if ( ! isset( $instance, $defaults ) ) {
	return wp_die( esc_html__( 'You cannot access this page directly', 'revision-strike' ) );
}

// Handle the execution of revisions.
if ( isset( $_GET['nonce'], $_POST['days'], $_POST['limit'] ) && wp_verify_nonce( $_GET['nonce'], 'revision-strike' ) ) {
	$args  = array(
		'days'  => absint( $_POST['days'] ),
		'limit' => absint( $_POST['limit'] ),
	);
	$instance->strike( $args );
	$stats = $instance->get_stats();
	$class = 'error';

	if ( 0 === $stats['count'] ) {
		$message = __( 'No revisions were found that matched your criteria.', 'revision-strike' );

	} elseif ( 0 === $stats['deleted'] && 0 < $stats['count'] ) {
		$message = __( 'Something went wrong deleting post revisions, please try again!', 'revision-strike' );

	} elseif ( 1 === $stats['deleted'] ) {
		$message = __( 'One post revision has been deleted successfully!', 'revision-strike' );
		$class   = 'updated';

	} else {
		/* Translators: %1$d represents the number of deleted revisions. */
		$message = sprintf( _n(
			'%1$d post revision has been deleted successfully!',
			'%1$d post revisions have been deleted successfully!',
			$stats['deleted'],
			'revision-strike'
		), $stats['deleted'] );
		$class   = 'updated';
	}

	printf(
		'<div class="%s"><p>%s</p></div>',
		esc_attr( $class ),
		esc_html( $message )
	);
}

?>

<div class="wrap">
	<div id="icon-tools" class="icon32"></div>
	<h2><?php esc_html_e( 'Revision Strike: On-Demand', 'revision-strike' ); ?></h2>

	<form method="POST" action="<?php echo esc_url( wp_nonce_url( 'tools.php?page=revision-strike', 'revision-strike', 'nonce' ) ); ?>">
		<p><?php
			esc_html_e(
				'This page will allow Revision Strike will remove old revisions from the post database immediately.',
				'revision-strike'
			);
		?></p>
		<p><?php
			/* Translators: %1$s is the URL of the "Writing" settings page. */
			echo wp_kses_post( sprintf( __(
				'To change the default settings for Revision Strike\'s automatic, daily run, please see <a href="%1$s">Settings &rsaquo; Writing</a>.',
				'revision_strike'
			), admin_url( 'options-writing.php' ) ) );
		?></p>

		<table class="form-table">
			<tr>
				<th scope="row"><?php esc_html_e( 'Post Age Threshold', 'revision-strike' ); ?></th>
				<td>
					<input name="days" type="number" class="small-text" value="<?php echo absint( $defaults['days'] ); ?>" />
					<p class="description"><?php
						esc_html_e(
							'A post must be published at least this many days before its revisions can be removed.',
							'revision-strike'
						);
					?></p>
				</td>
			</tr>
			<tr>
				<th scope="row"><?php esc_html_e( 'Limit', 'revision-strike' ); ?></th>
				<td>
					<input name="limit" type="number" class="small-text" value="<?php echo absint( $defaults['limit'] ); ?>" />
					<p class="description"><?php
						esc_html_e(
							'The maximum number of revisions to delete.',
							'revision-strike'
						);
					?></p>
				</td>
			</tr>
		</table>

		<?php submit_button( __( 'Strike Revisions', 'revision-strike' ), 'primary', null ); ?>
	</form>
</div>


revision-strike.php
<?php
/**
 * Plugin Name: Revision Strike
 * Plugin URI:  https://stevegrunwell.com/blog/revision-strike
 * Description: Periodically purge old post revisions via WP Cron.
 * Version:     0.6.0
 * Author:      Steve Grunwell
 * Author URI:  https://stevegrunwell.com
 * Text Domain: revision-strike
 * Domain Path: /languages
 *
 * @package Revision Strike
 * @author  Steve Grunwell
 */

require_once dirname( __FILE__ ) . '/includes/class-revision-strike.php';
require_once dirname( __FILE__ ) . '/includes/class-revision-strike-cli.php';
require_once dirname( __FILE__ ) . '/includes/class-settings.php';

/**
 * Bootstrap the plugin.
 */
function revisionstrike_init() {
	new RevisionStrike();

	load_plugin_textdomain( 'revision-strike', false, dirname( plugin_basename( __FILE__ ) ) . '/languages' );
}

add_action( 'init', 'revisionstrike_init' );

/**
 * Register the cron job on plugin activation.
 */
function revisionstrike_register_cron() {
	if ( false === wp_next_scheduled( RevisionStrike::STRIKE_ACTION ) ) {
		wp_schedule_event( time(), 'daily', RevisionStrike::STRIKE_ACTION );
	}
}

register_activation_hook( __FILE__, 'revisionstrike_register_cron' );

/**
 * Cancel the cron job when the plugin is disabled.
 */
function revisionstrike_deregister_cron() {
	wp_clear_scheduled_hook( RevisionStrike::STRIKE_ACTION );
}

register_deactivation_hook( __FILE__, 'revisionstrike_deregister_cron' );


includes/class-revision-strike-cli.php
<?php
/**
 * WP-CLI commands for Revision Strike.
 *
 * @package Revision Strike
 * @author Steve Grunwell
 */

if ( ! defined( 'WP_CLI' ) || ! WP_CLI ) {
	return;
}

/**
 * Remove revisions on old posts to clean up the database.
 */
class RevisionStrikeCLI extends WP_CLI {

	/**
	 * The current instance of the RevisionStrike class.
	 *
	 * @var RevisionStrike $instance
	 */
	protected $instance;

	/**
	 * Remove old post revisions.
	 *
	 * ## OPTIONS
	 *
	 * [--days=<days>]
	 * : Remove revisions on posts published at least <days> days ago. This is
	 * determined by the value set on Settings > Writing or a default of 30.
	 *
	 * [--limit=<limit>]
	 * : The maximum number of revisions to remove. This is determined by the
	 * value set on Settings > Writing or a default value of 50.
	 *
	 * [--post_type=<post_type>]
	 * : One or more post types (comma-separated) for which revisions should be
	 * struck. Default value is 'post'.
	 *
	 * [--verbose]
	 * : Enable verbose logging of deleted revisions.
	 *
	 * ## EXAMPLES
	 *
	 *   wp revision-strike clean
	 *   wp revision-strike clean --days=45
	 *   wp revision-strike clean --limit=75
	 *   wp revision-strike clean --post_type=post,page
	 *
	 * @synopsis [--days=<days>] [--limit=<limit>] [--post_type=<post_type>] [--verbose]
	 *
	 * @param array $args       A numeric array of position-based arguments.
	 * @param array $assoc_args An associative array of key-based arguments.
	 */
	public function clean( $args, $assoc_args ) {
		if ( isset( $assoc_args['verbose'] ) ) {
			add_action( 'wp_delete_post_revision', function ( $revision_id ) {
				WP_CLI::log( sprintf(
					/* Translators: %1$d represents the ID of the deleted revision. */
					esc_html__( 'Revision ID %1$d has been deleted.', 'revision-strike' ),
					$revision_id
				) );
			} );
		}

		$instance = $this->get_instance();
		$args     = array();

		foreach ( array( 'days', 'limit', 'post_type' ) as $arg ) {
			if ( isset( $assoc_args[ $arg ] ) ) {
				$args[ $arg ] = $assoc_args[ $arg ];
			}
		}

		$instance->strike( $args );

		WP_CLI::line();

		$stats = $instance->get_stats();
		if ( 0 === $stats['deleted'] ) {
			$message = __( 'No errors occurred, but no post revisions were removed.', 'revision-strike' );

		} elseif ( 1 === $stats['deleted'] ) {
			$message = __( 'One post revision was deleted successfully', 'revision-strike' );

		} else {
			/* Translators: %1$d represents the number of deleted revisions. */
			$message = sprintf( _n(
				'%1$d post revision was deleted successfully',
				'%1$d post revisions were deleted successfully',
				$stats['deleted'],
				'revision-strike'
			), $stats['deleted'] );
		}

		WP_CLI::success( esc_html( $message ) );
	}

	/**
	 * Remove *all* old post revisions from the WordPress database.
	 *
	 * This command will recursively strike all the old post revisions from the database,
	 * giving a site a fresh start that the regular Revision Strike cron job can maintain.
	 *
	 * ## OPTIONS
	 *
	 * [--days=<days>]
	 * : Remove revisions on posts published at least <days> days ago. This is
	 * determined by the value set on Settings > Writing or a default of 30.
	 *
	 * [--post_type=<post_type>]
	 * : One or more post types (comma-separated) for which revisions should be
	 * struck. Default value is 'post'.
	 *
	 * [--verbose]
	 * : Enable verbose logging of deleted revisions.
	 *
	 * ## EXAMPLES
	 *
	 *   wp revision-strike clean-all
	 *   wp revision-strike clean-all --days=45
	 *   wp revision-strike clean-all --post_type=post,page
	 *
	 * @synopsis [--days=<days>] [--post_type=<post_type>] [--verbose]
	 * @alias clean-all
	 *
	 * @param array $args       A numeric array of position-based arguments.
	 * @param array $assoc_args An associative array of key-based arguments.
	 */
	public function clean_all( $args, $assoc_args ) {
		$instance = $this->get_instance();
		$assoc_args['limit'] = $instance->count_eligible_revisions(
			isset( $assoc_args['days'] ) ? $assoc_args['days'] : $instance->settings->get_option( 'days' ),
			isset( $assoc_args['post_type'] ) ? $assoc_args['post_type'] : $instance->settings->get_option( 'post_type' )
		);

		return $this->clean( $args, $assoc_args );
	}

	/**
	 * Get the current RevisionStrike instance.
	 *
	 * @return RevisionStrike The current instance in $this->instance.
	 */
	protected function get_instance() {
		if ( null === $this->instance ) {
			$this->instance = new RevisionStrike();
		}
		return $this->instance;
	}
}

WP_CLI::add_command( 'revision-strike', 'RevisionStrikeCLI' );
